modules/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/task/TaskScheduler.java

```diff
@@ -0,0 +1,390 @@
+                    numberOfDaysToBeAdded = (int) dayOfWeek - startTime.get(Calendar.DAY_OF_WEEK);
+                } else if (startTime.get(Calendar.DAY_OF_WEEK) == (int) dayOfWeek && startTime.before(now)) {
+                    numberOfDaysToBeAdded = 7;
+                }
+            } else if (now.get(Calendar.MONTH) < (int) month) {
+                while (startTime.get(Calendar.MONTH) < (int) month) {
+                    if (startTime.get(Calendar.DAY_OF_WEEK) == (int) dayOfWeek) {
+                        startTime.add(Calendar.DATE, 7);
+                    } else if (startTime.get(Calendar.DAY_OF_WEEK) > (int) dayOfWeek) {
+                        startTime.add(Calendar.DATE, 7 - (startTime.get(Calendar.DAY_OF_WEEK) - (int) dayOfWeek));
+                    } else {
+                        startTime.add(Calendar.DATE, (int) dayOfWeek - startTime.get(Calendar.DAY_OF_WEEK));
+                    }
+                }
+            }
+            startTime.add(Calendar.DATE, numberOfDaysToBeAdded);
+        }
+        return startTime;
+    }
+
+    private static Calendar fixTheDateByDOM(Calendar now, Calendar startTime, long dayOfMonth) {
+        if (dayOfMonth >= 1) {
+            startTime.set(Calendar.DATE, (int) dayOfMonth);
+            if (startTime.before(now)) {
+                startTime.add(Calendar.MONTH, 1);
+            }
+        }
+        return startTime;
+    }
+
+    private static Calendar fixTheDateByMonth(Calendar now, Calendar startTime, long minute, long hour,
+                                              long dayOfWeek, long dayOfMonth, long month) {
+        if ((minute == 0 && hour > -1) || (minute == -1 && hour > -1)) {
+            startTime.set(Calendar.MINUTE, 0);
+            startTime.set(Calendar.SECOND, 0);
+            if (minute == -1) {
+                period = 59 * 60000;
+            }
+        } else if (minute > 0 && hour != -1) {
+            startTime.set(Calendar.SECOND, 0);
+        }
+        if (month > -1) {
+            if (startTime.get(Calendar.MONTH) < (int) month) {
+                startTime.add(Calendar.MONTH, (int) month - startTime.get(Calendar.MONTH));
+                if (dayOfWeek == -1 && dayOfMonth == -1) {
+                    startTime.set(Calendar.DATE, 1);
+                }
+            } else if (startTime.get(Calendar.MONTH) > (int) month) {
+                int months = (int) month - startTime.get(Calendar.MONTH);
+                startTime.add(Calendar.YEAR, 1);
+                startTime.add(Calendar.MONTH, months);
+                if (dayOfWeek == -1 && dayOfMonth == -1) {
+                    startTime.set(Calendar.DATE, 1);
+                }
+            } else if (startTime.get(Calendar.MONTH) == (int) month && startTime.before(now)) {
+                startTime.add(Calendar.YEAR, 1);
+            }
+        }
+
+        if (((now.get(Calendar.YEAR) < startTime.get(Calendar.YEAR)) || (
```

Comment:
Here also, it would be much cleaner to delegate this logic (logic inside the if condition) to a private method